### PR TITLE
Fix failed SCA authentication on subs. order payment not reflected in OFN

### LINKED ISSUE #7509: Failed authentication on Stripe-SCA payment leads to "paid" payment state
## Description

For subscription orders, when using test cards which require authentication (like x-3184), a link is sent per email to the customer, to provide for it. 

If authentication fails (clicking "Fail Authentication, in a test environment) or by using an invalid card, the UI displays the order to be complete and in a paid state; this info is showed both on the admin and customer accounts. The payment attempt appears as failed on the Stripe dashboard.

It also appears as paid in the UI but also on the DB:
```
irb(main):007:0> Spree::Order.where(number: 'R462515802')
=> #<ActiveRecord::Relation [#<Spree::Order id: 445537, number: "R462515802", item_total: 0.24e2, total: 0.64e2, state: "complete", adjustment_total: 0.4e2, user_id: 6927, created_at: "2021-04-29 15:52:38", updated_at: "2021-04-29 15:55:08", completed_at: "2021-04-29 15:52:40", bill_address_id: 102331, ship_address_id: 102332, payment_total: 0.64e2, shipment_state: "ready", payment_state: "paid", email: "filipefurtado+treec@gmail.com", special_instructions: nil, distributor_id: 1785, order_cycle_id: 3431, currency: "CAD", last_ip_address: nil, customer_id: 8152, created_by_id: nil, included_tax_total: 0.0, additional_tax_total: 0.0>]>
```

```
irb(main):010:0> Spree::Payment.where(order_id: 445537)
=> #<ActiveRecord::Relation [#<Spree::Payment id: 33640, amount: 0.64e2, order_id: 445537, created_at: "2021-04-29 15:52:39", updated_at: "2021-04-29 15:55:08", source_id: 4036, source_type: "Spree::CreditCard", payment_method_id: 751, state: "completed", response_code: "pi_1IlcB8KuuB1fWySnByQmNVb4", avs_response: nil, identifier: "KN7ZU2SA", cvv_response_code: nil, cvv_response_message: nil>]>
```

It appears as failed in Stripe:

## Expected Behavior

Payment state reflects the success/failure of the authentication step.

## Actual Behaviour

Payment state is displayed as "Paid", despite failed authentication.

## Steps to Reproduce

1. Set up x-3184 card for a given customer/hub combination.
2. Place a subscription for that customer/hub.
3. Manually fail the authentication step, when prompted to provide authentication
4. Check the order and payment states for both customer and hub.

## Workaround

None, this appears as paid an is not.

## Severity

S2 for investigation, not sure something similar is happening in production...

## Your Environment

* Version used: v3.6.9
* Browser name and version: Firefox 88
* Operating System and version (desktop or mobile): Ubuntu 20.04

### PR DESCRIPTION
**Pull request by sauloperez**
#### What? Why?

Closes #7509. This validates payment intents against Stripe and displays all the possible errors to the user.

This is how it looks:

Peek 2021-05-06 12-29

#### What should we test?

The exact scenario described in #7509.

#### Release notes

Fix Stripe's unauthorized payments not reflected in the order and show any possible error with the payment to the user.
Changelog Category: User facing changes

### CODE REVIEW

File: app/services/process_payment_intent.rb [hunk 1]
```diff
@@ -3,39 +3,63 @@
...
 
   def valid?
-    order.present? && valid_intent_string? && matches_last_payment?
+    order.present? && matches_last_payment?
   end
 
-  def valid_intent_string?
-    payment_intent&.starts_with?("pi_")
+  def validate_intent!
+    Stripe::PaymentIntentValidator.new.call(payment_intent, "stripe_account_id")
```
Comment (sauloperez) [inline app/services/process_payment_intent.rb:55]: `stripe_account_id` is what I need to figure out. I need to see where to pick that and how it's supposed to work.

**Comment by sauloperez** [comment]
I couldn't figure out how to ask Codecov to reassess the branch. That's clearly wrong.

**Comment by sauloperez** [comment]
:open_mouth: it looks good now. It may handle it eventually :shrug:

File: app/controllers/spree/orders_controller.rb [hunk 2]
```diff
@@ -26,8 +26,14 @@ class OrdersController < ::BaseController
...
-      ProcessPaymentIntent.new(params["payment_intent"], @order).call!
-      @order.reload
+
+      if params.key?("payment_intent")
+        result = ProcessPaymentIntent.new(params["payment_intent"], @order).call!
+        unless result.ok?
+          flash[:error] = "#{I18n.t("payment_could_not_process")}. #{result.error}"
+        end
+        @order.reload
+      end
```
Comment (sauloperez) [inline app/controllers/spree/orders_controller.rb:36]: I plan to move this to its own purpose-specific controller action in an upcoming PR so we don't mix the two concerns and potentially introduce bugs.

**Review by andrewpbrett — APPROVED** [review]
This looks excellent, thanks @sauloperez. The change that Matt mentions will be nice, as well as moving to a specific controller method. But that can come later 👍.

File: app/services/process_payment_intent.rb [hunk 3]
```diff
@@ -3,39 +3,72 @@
 # When directing a customer to Stripe to authorize the payment, we specify a
 # redirect_url that Stripe should return them to. When checking out, it's
 # /checkout; for admin payments and subscription payemnts it's the order url.
+#
```
Comment (sauloperez) [inline app/services/process_payment_intent.rb:6]: This change is crucial 😂 #OCD

File: app/controllers/spree/orders_controller.rb [hunk 4]
```diff
@@ -26,8 +26,14 @@ class OrdersController < ::BaseController
 
     def show
       @order = Spree::Order.find_by!(number: params[:id])
-      ProcessPaymentIntent.new(params["payment_intent"], @order).call!
-      @order.reload
+
+      if params.key?("payment_intent")
+        result = ProcessPaymentIntent.new(params["payment_intent"], @order).call!
+        unless result.ok?
+          flash[:error] = "#{I18n.t("payment_could_not_process")}. #{result.error}"
```
[the remark this answers is not in the document]
Comment (sauloperez) [reply app/controllers/spree/orders_controller.rb:33]: I have to check but I purposefully left the empty string as the default Result error message so this doesn't happen.

**Comment by filipefurtad0** [comment]
Hey @sauloperez ,

This looks good now. Did a brief sanity check, to see if authenticating the payment still leads to a complete and paid order:

Peek 2021-05-06 20-11

Looks good :+1:  After that, I reproduced the scenario leading to the bug, and now (as on the video above) it leads to the expected scenario:

- customer: the order appears as non paid
- admin: nothing changes

Peek 2021-05-06 19-58

If we do this the other way around - fail payment, then place a new sub with a valid payment - we get the flash message, but this is rather a previously observed "memory effect" on flash messages... It looks weird, but it's actually a separated issue:

Peek 2021-05-06 20-01

So, in summary, this is good to go :muscle:
Was great to pair on this! :raised_hands: 

The followup issues required (on my to-do list) are:
- No possibility for the admin to generate a new authentication link to the customer, after failed authentication (proposing bug-s3)
- The customer has a "green screen", despite failed payment: UI could be improved to display a more visible payment status (UI enhancement, it's not really a bug)
- "memory effect" on flash messages (proposing bug-s4)

**Comment by sauloperez** [comment]
> "memory effect" on flash messages (proposing bug-s4)

That is concerning :thinking: I think it deserves to be s3 because it's really misleading for the user.

File: app/controllers/spree/orders_controller.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (sauloperez) [reply app/controllers/spree/orders_controller.rb:33]: I see now. In that case `#ok?` returns `false` but `#error` is an empty string. As we don't pass a specific error message `flash[:error]` shows a generic `The payment could not be completed`. I left this open for further improvement. If we need to in the future, we can extract a [custom validation class](https://guides.rubyonrails.org/active_record_validations.html#custom-validators) and be more specific. For the purpose of this bugfix it felt too much.

**Comment by Matt-Yorkley** [comment]
> "memory effect" on flash messages (proposing bug-s4)

We had a similar issue in the past around checkout flash messages: https://github.com/openfoodfoundation/openfoodnetwork/pull/5906/files
